pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go

```diff
@@ -0,0 +1,112 @@
+)
+
+// TektonEventType holds the types of cloud events sent by Tekton
+type TektonEventType string
+
+const(
+  // TektonTaskRunUnknown is sent for TaskRuns with "ConditionSucceeded" "Unknown"
+  TektonTaskRunUnknown    TektonEventType = "dev.tekton.event.task.unknown"
+  // TektonTaskRunSuccessful is sent for TaskRuns with "ConditionSucceeded" "True"
+  TektonTaskRunSuccessful TektonEventType = "dev.tekton.event.task.successful"
+  // TektonTaskRunFailed is sent for TaskRuns with "ConditionSucceeded" "False"
+  TektonTaskRunFailed     TektonEventType = "dev.tekton.event.task.failed"
+)
+
+type ClientFactory func(...string) (client.Client, error)
+
+// SendCloudEvent sends a Cloud Event to the specified SinkURI
+func SendCloudEvent(sinkURI, eventID, eventSourceURI string, data []byte, eventType TektonEventType, logger *zap.SugaredLogger, cloudEventClientFactory ClientFactory) (cloudevents.Event, error) {
+	var event cloudevents.Event
+  // Setup the cloudevent client
+	cloudEventClient, err := cloudEventClientFactory(sinkURI)
+	if err != nil {
+		logger.Errorf("Error creating the cloud-event client: %s", err)
+    return event, err
+	}
+
+	cloudEventSource := types.ParseURLRef(eventSourceURI)
+	if cloudEventSource == nil {
+		logger.Errorf("Invalid eventSourceURI: %s", eventSourceURI)
+		return event, fmt.Errorf("Invalid eventSourceURI: %s", eventSourceURI)
+	}
+
+	event = cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			ID:         eventID,
+			Type:       string(eventType),
+			Source:     *cloudEventSource,
+			Extensions: nil,
+		}.AsV02(),
+		Data: data,
+	}
+	_, err = cloudEventClient.Send(context.TODO(), event)
+	if err != nil {
+		logger.Errorf("Error sending the cloud-event: %s", err)
+    return event, err
+	}
+  return event, nil
+}
+
+// SendTaskRunCloudEvent sends a cloud event for a TaskRun
+func SendTaskRunCloudEvent(sinkURI string, taskRun *v1alpha1.TaskRun, logger *zap.SugaredLogger, optionalClientFactory ...ClientFactory) (cloudevents.Event, error) {
+	// We take an optional clientFactory as input so we can mock it in tests
+	var cloudEventClientFactory ClientFactory
+	if len(optionalClientFactory) == 0 {
+		cloudEventClientFactory = kncloudevents.NewDefaultClient
+	} else {
+		cloudEventClientFactory = optionalClientFactory[0]
+	}
+	var event cloudevents.Event
+	// Check if the TaskRun is defined
+	if taskRun == nil {
+    return event, errors.New("Cannot send an event for an empty TaskRun")
+  }
+  eventID := taskRun.ObjectMeta.Name
+  taskRunStatus := taskRun.Status.GetCondition(apis.ConditionSucceeded)
+  var eventType TektonEventType
+  if taskRunStatus.IsUnknown() {
+    eventType = TektonTaskRunUnknown
+  } else if taskRunStatus.IsFalse() {
+    eventType = TektonTaskRunFailed
+  } else if taskRunStatus.IsTrue() {
+    eventType = TektonTaskRunSuccessful
+  } else {
+    return event, fmt.Errorf("Unknown condition for in TaskRun.Status %s", taskRunStatus.Status)
```

Comment:
I wonder if some event should be sent here.

Reply:
That `else` case is not really like to happen, but I wanted to at least log something in case someone crafted a broken TaskRun. I would not send an event though.

Reply:
OK